src/Microsoft.ML.CpuMath/AvxIntrinsics.cs

```diff
@@ -0,0 +1,1654 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// The exported function names need to be unique (can't be disambiguated based on signature), hence
+// we introduce suffix letters to indicate the general patterns used.
+// * A suffix means aligned and padded for SSE operations.
+// * U suffix means unaligned and unpadded.
+// * P suffix means sparse (unaligned) partial vector - the vector is only part of a larger sparse vector.
+// * Tran means the matrix is transposed.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.Intrinsics;
+using System.Runtime.Intrinsics.X86;
+
+namespace Microsoft.ML.Runtime.Internal.CpuMath
+{
+    public static class AvxIntrinsics
+    {
+        private const int Vector256Alignment = 32;
+
+        private static bool Compat(AlignedArray a)
+        {
+            Contracts.AssertValue(a);
+            Contracts.Assert(a.Size > 0);
+            return a.CbAlign == Vector256Alignment;
+        }
+
+        private static unsafe float* Ptr(AlignedArray a, float* p)
+        {
+            Contracts.AssertValue(a);
+            float* q = p + a.GetBase((long)p);
+            Contracts.Assert(((long)q & (Vector256Alignment - 1)) == 0);
+            return q;
+        }
+
+        [MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
+        private static Vector256<float> ToVector256(in Vector128<float> a, in Vector128<float> b)
+        {
+            // REVIEW NEEDED: Is it the correct port of the following code?
+            // #ifndef _WIN32
+            // #define _mm256_set_m128(va, vb) _mm256_insertf128_ps(_mm256_castps128_ps256(vb), va, 1)
+            // #endif
+            return Avx.InsertVector128(Avx.ExtendToVector256(b), a, 1);
+        }
+
+        [MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
+        private static void ZeroUpper()
+        {
+            // Currently no-op since _mm256_zeroupper is not supported (ref: https://github.com/dotnet/coreclr/pull/16955)
+            // This is a placeholder in case the intrinsic is supported later on.
+            return;
+        }
+
+        [MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
+        private static Vector128<float> GetLow(in Vector256<float> x)
+        {
+            return Avx.ExtractVector128(x, 0);
```

Comment:
Another option to implement GetLow is to use `Avx.GetLowerHalf`, which has no code gen. So that one instruction could be saved.

Reply:
Thank you!  Would `GetHigh` below also benefit from something like `Avx.GetHigherHalf`?

Reply:
We do not have `Avx.GetHigherHalf` intrinsic, just use `Avx.ExtractVector128(x, 1);` 😄